qpext/cmd/qpext/main.go

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2022 The KServe Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/go-multierror"
+	logger "github.com/kserve/kserve/qpext"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/expfmt"
+	"go.uber.org/zap"
+	"io"
+	"mime"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+import "knative.dev/serving/pkg/queue/sharedmain"
+
+var (
+	promRegistry *prometheus.Registry
+)
+
+const (
+	// aggregate scraping env vars from kserve/pkg/constants
+	KServeContainerPrometheusMetricsPortEnvVarKey     = "KSERVE_CONTAINER_PROMETHEUS_METRICS_PORT"
+	KServeContainerPrometheusMetricsPathEnvVarKey     = "KSERVE_CONTAINER_PROMETHEUS_METRICS_PATH"
+	QueueProxyAggregatePrometheusMetricsPortEnvVarKey = "AGGREGATE_PROMETHEUS_METRICS_PORT"
+	QueueProxyMetricsPort                             = "9091"
+	DefaultQueueProxyMetricsPath                      = "/metrics"
+)
+
+type ScrapeConfigurations struct {
+	logger         *zap.Logger
+	QueueProxyPath string `json:"path"`
+	QueueProxyPort string `json:"port"`
+	AppPort        string
+	AppPath        string
+}
+
+func getURL(port string, path string) string {
+	return fmt.Sprintf("http://localhost:%s%s", port, path)
+}
+
+// getHeaderTimeout parse a string like (1.234) representing number of seconds
+func getHeaderTimeout(timeout string) (time.Duration, error) {
+	timeoutSeconds, err := strconv.ParseFloat(timeout, 64)
+	if err != nil {
+		return 0 * time.Second, err
+	}
+
+	return time.Duration(timeoutSeconds * 1e9), nil
+}
+
+func applyHeaders(into http.Header, from http.Header, keys ...string) {
+	for _, key := range keys {
+		val := from.Get(key)
+		if val != "" {
+			into.Set(key, val)
+		}
+	}
+}
+
+func negotiateMetricsFormat(contentType string) expfmt.Format {
+	mediaType, _, err := mime.ParseMediaType(contentType)
+	if err == nil && mediaType == expfmt.OpenMetricsType {
+		return expfmt.FmtOpenMetrics
+	}
+	return expfmt.FmtText
+}
+
+// scrapeAndWriteAgentMetrics gathers a slice of prometheus metric families and encodes the metrics.
+func scrapeAndWriteAgentMetrics(w io.Writer) error {
+	mfs, err := promRegistry.Gather()
+	enc := expfmt.NewEncoder(w, expfmt.FmtText)
+	if err != nil {
+		return err
+	}
```

Comment:
Check the error before doing more stuff
```suggestion
	if err != nil {
		return err
	}
	enc := expfmt.NewEncoder(w, expfmt.FmtText)
```